### ISSUE TITLE
MIgrate off of PostgreSQL

### ISSUE THREAD

**Issue by guseggert** [libp2p/hydra-booster#153]
We have reached the point where scaling the current Postgres setup is taking too much effort, so we have agreed to store the provider records in a hosted key-value store instead, to reduce the maintenance burden of operating hydras by enabling the backing store to scale horizontally.

The current consensus is to move to DynamoDB, due to my familiarity with it and the relative urgency.

## Table Design
- key (bytes)
  - Primary Key
  - typically this is the raw multihash of a CID
- ttl (number)
  - Sort Key
  - unix epoch timestamp of when the entry expires
  - configure your DynamoDB table to use this as an item TTL for auto-eviction
  - since this is a sort key, results for a given CID are ordered by this, so that you can return the most recently-cached providers first
- prov (bytes)
  - the raw peer ID of the provider of the given key

A DHT put results in a single PutItem request to DynamoDB.

A DHT query results in a paginated DynamoDB Query, keyed on the CID multihash. The query returns an item per provider of the CID, up to a pre-configured limit. For popular CIDs with many providers, this could take many round-trips to DynamoDB to fetch subsequent pages. A single page can return up to 1 MB of items. Since items are automatically evicted by DynamoDB based on their TTL, this automatically excludes expired entries.

## Todo
- [x] Add details here about Postgres issues
- [x] Re-enable proactively finding providers
- [x] Decide on AWS account
- [x] Define infra as code (no making tables by hand)
- [x] Ensure table uses TTL for cache eviction
- [x] Determine if possible to have persistent peer ids
- [x] Scale up
- [x] Add gratuitous metrics in codebase
  - [x] # pages read in query response
  - [x] provider distribution per cid
  - [x] # truncated queries
  - [x] basic request metrics (# reqs, # success, # errors, latency, # retries, etc) 
- [x] Update README/docs describing motivation and background of this decision
- [x] Update dashboards
- [x] Add some basic runbook about what errors we may encounter and how to mitigate 
  - What do I do if DDB is throttling?
  - What do I do if hosts are running out of memory or CPU?
  - What do I do if there is high latency?
- [ ] Persist the IPNS records (can reevaluate if this is a pain.  Timebox to a day.)
- [x] Runbook easily discoverable (move to external repo)
- [x] Retire the old stuff
- [x] Link the various PRs for this project
- [x] Other repo cleanup work concerning Alan repos and postgres datastore.

**Comment by guseggert** [comment]
I've got the basic impl in this branch, which I've been load testing: https://github.com/libp2p/hydra-booster/tree/ddb

I want to add some DB-level metrics prior to deploying the MVP. After deploying I will also write some more tests (it does have some basic tests but they aren't rigorous). This also needs some infra work to pick an account, create the table, and wire it up. @thattommyhall was considering moving the entire k8s cluster to AWS, I"m not sure of the status of that. It's also not strictly neccessary to mitigate this issue, just a nice-to-have. If we don't move th

**Comment by BigLep** [comment]
Thanks @guseggert for putting this together.  Feel free to resolve https://github.com/protocol/hydra-booster-k8s/issues/2 if it feels like a duplicate.

**Comment by guseggert** [comment]
Update: I've been running these for most of the week and tweaking configuration since we are running on different hosts now. Everything looks good so far.

I have set up a public operational dashboard here: https://cloudwatch.amazonaws.com/dashboard.html?dashboard=Hydras&context=eyJSIjoidXMtZWFzdC0xIiwiRCI6ImN3LWRiLTY4NjMxMTAxMzcxMCIsIlUiOiJ1cy1lYXN0LTFfQnZIM2ZLSWdUIiwiQyI6IjZobm9mOHBiaWd0ZWFxN2VzdnZqMnVxa2xoIiwiSSI6InVzLWVhc3QtMTpkNzdkNWFjMi1mY2RkLTRmOWItOTFiNC1mMmNkZjRkNTQ2NDMiLCJNIjoiUHVibGljIn0%3D#dashboards:

@thattommyhall is working on some final Terraform changes to give the nodes stable PeerIDs (currently they get new ids every time they restart, which causes some network churn). This effectively pegs the cluster at a certain size, preventing us from autoscaling, which isn't ideal, but we don't have a solid metric to use for autoscaling anyway as the ideal # heads is really a function of the holistic network and not of any localized metrics, and we are lacking those (this is related to https://github.com/libp2p/hydra-booster/issues/93). 

There were a few issues with the Postgres DB:

* The DB ran out of disk space and was placed into read-only mode, and we did not detect this due to lack of monitoring
* Once we added disk space, we did not have enough conns in the conn pool on each node, causing conn contention and goroutine pileups due to spawning unbounded goroutines, eventually leading to OOMing
* Once we added more connections on each host, the DB was unable to handle the load due to lock contention on the writes
* We tweaked the write query to batch transactions which caused IO contention, then we lowered batch sizes and found a happy batch size
* The autovacuumer was unable to keep up with the amount of transactions happening, causing txid exhaustion, we worked with the hosting provider to find stable settings

**Comment by guseggert** [comment]
Yesterday I scaled up the fleet to 67 hosts w/ 15 heads per host, giving us 1005 heads on the new fleet. We're letting it run over the weekend before we decommission the old fleet, to make sure it reaches a comparable steady state.

I've added a runbook here, which is currently private: https://github.com/protocol/bifrost-infra/blob/add-hydras/terraform/deployments/hydra-boosters/RUNBOOK.md

@thattommyhall has plans to rip out the hydras infra from that repo and make it public.

Currently I'm updating the dashboard (and fixing some existing graphs that are buggy).

**Comment by guseggert** [comment]
Here's a (still-private) PR containing the commits on the infra side for this: https://github.com/protocol/bifrost-infra/pull/1584

**Comment by guseggert** [comment]
Re: persisting IPNS records, I've written a DynamoDB datastore here: https://github.com/guseggert/go-ds-dynamodb

I'm currently running it locally. It works great for the peerstore, but I have yet to see any DHT entries in it, because my node hasn't received any PutValue reqs. I'll wait around on those for a bit, and if I don't get any, will take matters into my own hands.

**Comment by BigLep** [comment]
2022-02-10: 
- Waiting on Tom for some repo consolidation work.
- Need to do some repo cleanup work

**Comment by BigLep** [comment]
2022-03-03 update: load testing "Persist the IPNS records"
* Going to take out peerstore.
* Will deploy today.

**Comment by guseggert** [comment]
IPNS PR: https://github.com/libp2p/hydra-booster/pull/161

It is already deployed (I load tested, then ran on a one-box, then deployed to the whole fleet). I also added infra code for a "one box" for deploying changes to a single host before the entire fleet, for testing new changes.

I also moved the random seed that we use for peer ID generation into Secrets Manager, which was the last remaining thing to do before making the infra code public. I've created the repo here: https://github.com/libp2p/hydra-booster-infra/

I am waiting on @thattommyhall as a second set of eyes to make sure I haven't missed removing any secrets, and then I'll copy the code over
